### PR TITLE
Startup Wizard - Addition of Startup Wizard

### LINKED ISSUE #269: Add setup wizard
First time users need to setup the jellyfin server by adding a user and setting up the libraries.

### PR DESCRIPTION
**Pull request by Camc314**
- Closes #269 
 - Left to do. Add ability to add and configure libraries in the wizard

### CODE REVIEW

**Comment by MrTimscampi** [comment]
> Add ability to add and configure libraries in the wizard

For this step, keep in mind the mid-term plan is to remove that step from the wizard and replace it with 2 steps of plugin installs: one for features and one for providers.

So I'm not sure if it's worth spending much time on adding that here :)

File: components/wizard/WizardAdminAccount.vue [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      :type="showPassword ? 'text' : 'password'"
+      @click:append="() => (showPassword = !showPassword)"
+    ></v-text-field>
+    <v-text-field
+      v-model="passwordCheck"
+      outlined
+      :label="$t('confirmPassword')"
+      required
+      type="password"
+      :rules="[(v) => (!!v && v === admin.Password) || $t('bothPasswordsSame')]"
```
Comment (dkanada) [inline components/wizard/WizardAdminAccount.vue:26]: cc @MrTimscampi are you always adding the rules inline? I've found moving them to an object in the component to be much more managable over time, especially when you want more than one per item. You can also create a utility somewhere to handle common rules and add them wherever you want to reduce code duplication, which helps for more complex rules.

File: locales/en-US.json [hunk 2]
```diff
@@ -41,11 +44,14 @@
...
   "signIn": "Sign in",
+  "userSuccessfullySet": "User successfully set.",
   "unexpectedError": "Unexpected error",
   "unhandledException": "Unhandled exception",
   "upNext": "Up next",
+  "username": "Username",
   "filtersNotFound": "Unable to load filters",
   "status": "Status",
+  "passwordRequired": "Password Required",
   "preferedLanguage": "Prefered language",
```
Comment (dkanada) [inline locales/en-US.json]: stick to one casing method for titles :p

File: components/wizard/WizardAdminAccount.vue (hunk 1 above)
Comment (MrTimscampi) [reply components/wizard/WizardAdminAccount.vue:26]: For now, we do a combination of both.

I've been thinking about moving validation to something like [VeeValidate](https://github.com/logaretm/vee-validate), though.

**Comment by Camc314** [comment]
> > Add ability to add and configure libraries in the wizard
> 
> For this step, keep in mind the mid-term plan is to remove that step from the wizard and replace it with 2 steps of plugin installs: one for features and one for providers.
> 
> So I'm not sure if it's worth spending much time on adding that here :)

Wouldn't it be better to have the add libraries step, then based on the type of libraries have a step for adding the relevant plugins (e.g. MusicBrainz for music...)

File: locales/en-US.json (hunk 2 above)
Comment (Camc314) [reply locales/en-US.json]: Bad habits :p

**Comment by MrTimscampi** [comment]
We want the providers to be there before the library creation (otherwise scan starts and the user has to go edit the libraries) and the server will require a restart to load the plugins.

So the idea is to install the plugins, then restart and, finally, gently onboard them towards adding libraries.

**Comment by MrTimscampi** [comment]
@Camc314 What's the status on this? Could we merge it without the library creation, then improve it later?

**Comment by MrTimscampi** [comment]
Brand new server, I seem to be unable to create the admin account, so I can't get further than that for testing.

Some early things:

* I'm not sure how this will play with the server, but we should probably be able to jump back to previous steps.
* The forms are way too large. Something like the login/server pages would feel better
* The "Previous" button should be the `secondary` color
* I'm not sure the page descriptions like "This account will be used for managing the server" are necessary
* The "confirm password" field is missing the visibility toggle
* Translation keys should be nested. A main key `firstTimeWizard` with sub-keys per step or something would fit and keep it more organized
* "Language and Locale" uses title case
* The first step is missing the `Dates and numbers` settings (It's different from languages, so one can't be the other. Plus people might want a language with a specific locale, like English (US) for language and German for dates and times)

The traceback I get on the admin account is the following:
```
[2021-01-05 23:30:21.064 +01:00] [ERR] [105] Jellyfin.Server.Middleware.ExceptionMiddleware: Error processing request. URL "POST" "/Startup/User".
System.InvalidOperationException: Sequence contains no elements
   at System.Linq.ThrowHelper.ThrowNoElementsException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Jellyfin.Api.Controllers.StartupController.UpdateStartupUser(StartupUserDto startupUserDto) in /tmp/makepkg/jellyfin-git/src/jellyfin/Jellyfin.Api/Controllers/StartupCont
roller.cs:line 133
   at lambda_method64(Closure , Object )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Jellyfin.Server.Middleware.ServerStartupMessageMiddleware.Invoke(HttpContext httpContext, IServerApplicationHost serverApplicationHost, ILocalizationManager localizationManager) in /tmp/makepkg/jellyfin-git/src/jellyfin/Jellyfin.Server/Middleware/ServerStartupMessageMiddleware.cs:line 49
   at Jellyfin.Server.Middleware.WebSocketHandlerMiddleware.Invoke(HttpContext httpContext, IWebSocketManager webSocketManager) in /tmp/makepkg/jellyfin-git/src/jellyfin/Jellyfin.Server/Middleware/WebSocketHandlerMiddleware.cs:line 38
   at Jellyfin.Server.Middleware.IpBasedAccessValidationMiddleware.Invoke(HttpContext httpContext, INetworkManager networkManager, IServerConfigurationManager serverConfigurationManager) in /tmp/makepkg/jellyfin-git/src/jellyfin/Jellyfin.Server/Middleware/IpBasedAccessValidationMiddleware.cs:line 75
   at Jellyfin.Server.Middleware.LanFilteringMiddleware.Invoke(HttpContext httpContext, INetworkManager networkManager, IServerConfigurationManager serverConfigurationManager) in /tmp/makepkg/jellyfin-git/src/jellyfin/Jellyfin.Server/Middleware/LanFilteringMiddleware.cs:line 46
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Swashbuckle.AspNetCore.ReDoc.ReDocMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Jellyfin.Server.Middleware.RobotsRedirectionMiddleware.Invoke(HttpContext httpContext) in /tmp/makepkg/jellyfin-git/src/jellyfin/Jellyfin.Server/Middleware/RobotsRedirectionMiddleware.cs:line 45
   at Jellyfin.Server.Middleware.LegacyEmbyRouteRewriteMiddleware.Invoke(HttpContext httpContext) in /tmp/makepkg/jellyfin-git/src/jellyfin/Jellyfin.Server/Middleware/LegacyEmbyRouteRewriteMiddleware.cs:line 52
   at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.Invoke(HttpContext context)
   at Jellyfin.Server.Middleware.ResponseTimeMiddleware.Invoke(HttpContext context) in /tmp/makepkg/jellyfin-git/src/jellyfin/Jellyfin.Server/Middleware/ResponseTimeMiddleware.cs:line 64
   at Jellyfin.Server.Middleware.ExceptionMiddleware.Invoke(HttpContext context) in /tmp/makepkg/jellyfin-git/src/jellyfin/Jellyfin.Server/Middleware/ExceptionMiddleware.cs:line 101
```

**Comment by Camc314** [comment]
> * I'm not sure how this will play with the server, but we should probably be able to jump back to previous steps.
>   Edit: I notice we can with the Previous button. Being able to click on the stepper at the top would be nice as well.

The only problem I see with this is if someone were to jump ahead to the final step, complete the wizard without completing the prior steps. This could cause issues as there may be no admin account

> The forms are way too large. Something like the login/server pages would feel better

The purpose of it being larger is so that the text above is visible:
I think it is better this way as it means the names of all steps can be seen
image


> The "Previous" button should be the secondary color

This makes the text even darker
Secondary:
image
Undefined:
image

> I'm not sure the page descriptions like "This account will be used for managing the server" are necessary

Removed

> The "confirm password" field is missing the visibility toggle

Added

> Translation keys should be nested. A main key firstTimeWizard with sub-keys per step or something would fit and keep it more organized


> "Language and Locale" uses title case

updated

> The first step is missing the Dates and numbers settings (It's different from languages, so one can't be the other. Plus people might want a language with a specific locale, like English (US) for language and German for dates and times)

Where are you getting the Dates and numbers setting from? As far as I can see it is not pointed out in the startupwizard API endpoints.

> The buttons should be centered and forced to the same size (like what's done on the login/select user/server pages)

With making the v-stepper wider than before, I think having the buttons size not take up the full width

> The traceback I get on the admin account is the following:

I think this may be due to the device ID changing? I have moved setting the device profile to the wizard page instead of being set everytime the account is created/update.

**Comment by MrTimscampi** [comment]
> The only problem I see with this is if someone were to jump ahead to the final step, complete the wizard without completing the prior steps. This could cause issues as there may be no admin account

Adding `editable` to `v-stepper-step` should make previous steps clickable without allowing to skip any of them (https://vuetifyjs.com/en/api/v-stepper-step/#props)

> The purpose of it being larger is so that the text above is visible:
I think it is better this way as it means the names of all steps can be seen

Agreed, but couldn't we have two width for these? The horizontal one won't work on mobile anyway, so having it become vertical on, say, `xs` and `sm` maybe, but horizontal on larger screens would work here. Then have the form itself act like the others but the stepper stay a larger width could work, I think.

But this is not blocking at all, we can look at it in further PRs, so no worries.

> This makes the text even darker

I was thinking of removing the `text` prop to make it a normal button and have it use the secondary color. We don't really use text buttons for anything elsewhere, so the normal ones may fit better.

> Where are you getting the Dates and numbers setting from? As far as I can see it is not pointed out in the startupwizard API endpoints.

Hm, I was thing we had that in the startup, but I might have been mistaken. Seems like the setting I'm thinking about is client-side only on old web.

Not needed now, but a potential future improvement for this would be to mirror the selected language here to the user preferences for the client's display language (And same for the locale).

> With making the v-stepper wider than before, I think having the buttons size not take up the full width

For sure, but with the narrower form (keeping the stepper as-is), this would be doable and would look more in sync with the other pages.

Again, not blocking though, we can mess with that at a later date.

> I think this may be due to the device ID changing? I have moved setting the device profile to the wizard page instead of being set everytime the account is created/update.

Thanks, I'll re-test :)

**Comment by MrTimscampi** [comment]
The wizard works perfectly now :+1: 

Gonna review the code later today, once I have some free time :smiley: 

One issue I could see during the test is the `allowRemoteAccess` key doesn't have a translation.

**Comment by Camc314** [comment]
I've fixed making the steps editable. The `editable` prop allows skipping steps, so it is now computed, if a step should be editable or not.

> I was thinking of removing the `text` prop to make it a normal button and have it use the secondary color. We don't really use text buttons for anything elsewhere, so the normal ones may fit better.
> 

Updated, it looks better now the text prop is gone

> Agreed, but couldn't we have two width for these? The horizontal one won't work on mobile anyway, so having it become vertical on, say, xs and sm maybe, but horizontal on larger screens would work here. Then have the form itself act like the others but the stepper stay a larger width could work, I think.

On smaller screens, the v-stepper hides the section titles

image


Ah I see what you mean, without dev tools open it is very wide lol. Should be a bit better now.

File: components/wizard/WizardMetadata.vue [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      try {
+        await this.$api.startup.updateInitialConfiguration({
+          startupConfigurationDto: {
+            ...this.initialConfig,
+            MetadataCountryCode: this.metadataLanguage,
+            PreferredMetadataLanguage: this.metadataCountry
+          }
+        });
+
+        this.pushSnackbarMessage({
```
Comment (MrTimscampi) [inline components/wizard/WizardMetadata.vue:74]: Same here.

File: components/wizard/WizardAdminAccount.vue [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      :label="$t('username')"
+      type="username"
+      required
+    ></v-text-field>
+    <v-text-field
+      v-model="admin.Password"
+      outlined
+      :label="$t('password')"
+      :rules="[(v) => !!v || $t('passwordRequired')]"
+      required
```
Comment (MrTimscampi) [inline components/wizard/WizardAdminAccount.vue:15]: Passwords aren't required here in the current web client.

File: locales/en-US.json [hunk 5]
```diff
@@ -240,6 +246,14 @@
   "specialFeatures": "Special Features",
   "status": "Status",
   "studios": "Studios",
+  "username": "Username",
+  "filtersNotFound": "Unable to load filters",
+  "finish": "Finish",
+  "passwordRequired": "Password Required",
```
Comment (MrTimscampi) [inline locales/en-US.json:252]: Some of these aren't needed with VeeValidate.

File: pages/wizard.vue [hunk 6]
```diff
@@ -0,0 +1,155 @@
+<template>
+  <v-container fill-height>
+    <v-row align="center" justify="center">
+      <v-col cols="12" sm="12" md="12" xl="8">
+        <h1 class="text-h4 mb-6 text-center">{{ heading }}</h1>
```
Comment (MrTimscampi) [inline pages/wizard.vue:5]: Not sure if we should have a header here.

Technically, the stepper already has the title. By making the stepper vertical on small screen, we can keep the headers visible and not need this.

File: pages/wizard.vue [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            <v-divider></v-divider>
+
+            <v-stepper-step
+              :complete="wizardStage > 3"
+              step="3"
+              :editable="maxWizardStage > 2"
+            >
+              {{ $t('wizard.preferredMetadataLanguage') }}
+            </v-stepper-step>
+            <v-divider></v-divider>
```
Comment (MrTimscampi) [inline pages/wizard.vue:35]:
```suggestion
            </v-stepper-step>

            <v-divider></v-divider>
```

**Review by MrTimscampi — CHANGES_REQUESTED** [review]

File: components/wizard/WizardMetadata.vue [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+      metadataLanguage: '',
+      metadataCountry: '',
+      initialConfig: {} as StartupConfigurationDto,
+      cultureOptions: [] as CultureDto[],
+      countryOptions: [] as CountryInfo[],
+      loading: false
+    };
+  },
+  async created() {
+    this.initialConfig = (
```
[the remark this answers is not in the document]
Comment (Camc314) [reply components/wizard/WizardMetadata.vue:51]: The only issue is that it changes in several steps, I found it easier to fetch each time a step is reached

File: locales/en-US.json (hunk 5 above)
Comment (Camc314) [reply locales/en-US.json]: I've updated it to use vee-validate

File: components/wizard/WizardAdminAccount.vue (hunk 4 above)
Comment (Camc314) [reply components/wizard/WizardAdminAccount.vue]: Imho we should make the admin account have a password. Regardless, I have removed it.

File: components/wizard/WizardAdminAccount.vue [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        });
+
+        this.$emit('step-complete', { step: 2 });
+      } catch (error) {
+        // eslint-disable-next-line no-console
+        console.error(error);
+        this.pushSnackbarMessage({
+          message: this.$t('wizard.setAdminError'),
+          color: 'error'
+        });
```
[the remark this answers is not in the document]
Comment (Camc314) [reply components/wizard/WizardAdminAccount.vue:91]: That is a good idea

File: components/wizard/WizardAdminAccount.vue [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      outlined
+      :label="$t('username')"
+      type="username"
+      required
+    ></v-text-field>
+    <v-text-field
+      v-model="admin.Password"
+      outlined
+      :label="$t('password')"
+      :rules="[(v) => !!v || $t('passwordRequired')]"
```
[the remark this answers is not in the document]
Comment (Camc314) [reply components/wizard/WizardAdminAccount.vue]: I've updated it to use vee-validate

File: plugins/veeValidate.ts [hunk 11]
```diff
@@ -9,6 +9,15 @@ extend('mustBeUrl', (value: string): boolean => {
   return /^https?:\/\/.+/.test(value);
 });
 
+extend('bothPasswordsSame', {
+  params: ['target'],
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  validate(value, { target }) {
+    return value === target;
```
Comment (Camc314) [inline plugins/veeValidate.ts:17]: This ts-comment is not ideal, but this is what the docs show to do to ensure that passwords are equal

https://logaretm.github.io/vee-validate/advanced/cross-field-validation.html#targeting-other-fields

**Review by MrTimscampi — APPROVED** [review]
LGTM :rocket:

File: plugins/veeValidate.ts (hunk 11 above)
Comment (ferferga) [reply plugins/veeValidate.ts:17]: We should have added a comment there with that before merging

**Pull request merged**